iroh-bytes/src/baomap.rs

```diff
@@ -0,0 +1,288 @@
+        target: PathBuf,
+        mode: ExportMode,
+        progress: impl Fn(u64) -> io::Result<()> + Send + Sync + 'static,
+    ) -> BoxFuture<'_, io::Result<()>>;
+}
+
+/// The mutable part of a BaoDb
+pub trait Store: ReadonlyStore + PartialMap {
+    /// This trait method imports a file from a local path.
+    ///
+    /// `data` is the path to the file.
+    /// `mode` is a hint how the file should be imported.
+    /// `progress` is a sender that provides a way for the importer to send progress messages
+    /// when importing large files. This also serves as a way to cancel the import. If the
+    /// consumer of the progress messages is dropped, subsequent attempts to send progress
+    /// will fail.
+    ///
+    /// Returns the hash of the imported file. The reason to have this method is that some database
+    /// implementations might be able to import a file without copying it.
+    fn import(
+        &self,
+        data: PathBuf,
+        mode: ImportMode,
+        progress: impl ProgressSender<Msg = ImportProgress> + IdGenerator,
+    ) -> BoxFuture<'_, io::Result<(Hash, u64)>>;
+
+    /// This trait method imports data from memory.
+    ///
+    /// It is a special case of `import` that does not use the file system.
+    fn import_bytes(&self, bytes: Bytes) -> BoxFuture<'_, io::Result<Hash>>;
+}
+
+/// Progress messages for an import operation
+///
+/// An import operation involves computing the outboard of a file, and then
+/// either copying or moving the file into the database.
+#[allow(missing_docs)]
+#[derive(Debug)]
+pub enum ImportProgress {
+    /// Found a path
+    ///
+    /// This will be the first message for an id
+    Found { id: u64, path: PathBuf },
+    /// Progress when copying the file to the store
+    ///
+    /// This will be omitted if the store can use the file in place
+    ///
+    /// There will be multiple of these messages for an id
+    CopyProgress { id: u64, offset: u64 },
+    /// Determined the size
+    ///
+    /// This will come after `Found` and zero or more `CopyProgress` messages.
+    /// For unstable files, determining the size will only be done once the file
+    /// is fully copied.
+    Size { id: u64, size: u64 },
+    /// Progress when computing the outboard
+    ///
+    /// There will be multiple of these messages for an id
+    OutboardProgress { id: u64, offset: u64 },
+    /// Done computing the outboard
+    ///
+    /// This comes after `Size` and zero or more `OutboardProgress` messages
+    OutboardDone { id: u64, hash: Hash },
+}
+
+/// The import mode describes how files will be imported.
+///
+/// This is a hint to the import trait method. For some implementations, this
+/// does not make any sense. E.g. an in memory implementation will always have
+/// to copy the file into memory. Also, a disk based implementation might choose
+/// to copy small files even if the mode is `Reference`.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
+pub enum ImportMode {
```

Comment:
I guess the plan is to add `Reflink` as an option here when we have support for it?

Reply:
Yes. But also if you have 2 options an enum is more clear than a flag. For the flag you never know if true means on or off.

Not sure how to deal with reflink: have a separate option or just treat it as a fast way to copy?